pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go

```diff
@@ -0,0 +1,243 @@
+	event *roachpb.RangeFeedEvent
+	err   error
+}
+
+// channelRangeFeedEventProducer is a rangeFeedEventProducer which receives
+// events on input channel, and returns events when Recv is called.
+type channelRangeFeedEventProducer struct {
+	ctx   context.Context
+	errCh chan error // Signalled to propagate terminal error the consumer.
+	input chan eventOrError
+}
+
+// Recv implements rangeFeedEventProducer interface.
+func (c *channelRangeFeedEventProducer) Recv() (*roachpb.RangeFeedEvent, error) {
+	select {
+	case <-c.ctx.Done():
+		return nil, c.ctx.Err()
+	case err := <-c.errCh:
+		return nil, err
+	case e := <-c.input:
+		return e.event, e.err
+	}
+}
+
+var _ rangeFeedEventProducer = (*channelRangeFeedEventProducer)(nil)
+
+// startMuxRangeFeed begins the execution of rangefeed for the specified RangeFeedRequest.
+func (m *muxer) startMuxRangeFeed(
+	ctx context.Context, rpc rpcClient, req *roachpb.RangeFeedRequest,
+) (rangeFeedEventProducer, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if m.mu.terminalErr != nil {
+		return nil, m.mu.terminalErr
+	}
+
+	// Lookup or establish MuxRangeFeed for this node.
+	nodeID := req.Replica.NodeID
+	stream, found := m.mu.muxClients[nodeID]
+	if !found {
+		var err error
+		stream, err = rpc.InternalClient.MuxRangeFeed(rpc.ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		m.mu.muxClients[req.Replica.NodeID] = stream
+		m.g.GoCtx(func(ctx context.Context) error {
+			return m.receiveEventsFromNode(ctx, nodeID, stream)
+		})
+	}
+
+	streamID := m.mu.nextStreamID
+	m.mu.nextStreamID++
+
+	// Start RangeFeed for this request.
+	req.StreamID = streamID
+	if err := stream.Send(req); err != nil {
+		return nil, err
+	}
+	m.mu.activeFeeds[nodeID] = append(m.mu.activeFeeds[nodeID], streamID)
+	producer := &channelRangeFeedEventProducer{
+		ctx:   ctx,
+		errCh: make(chan error, 1),
+		input: make(chan eventOrError),
+	}
+	m.mu.producers[streamID] = producer
+	return producer, nil
+}
+
+// demux de-multiplexes events and sends them to appropriate rangefeed event consumer.
+func (m *muxer) demux(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case e := <-m.eventCh:
+			m.mu.Lock()
+			producer, found := m.mu.producers[e.StreamID]
+			m.mu.Unlock()
+
+			if !found {
+				return m.shutdownWithError(errors.AssertionFailedf(
+					"expected to find consumer for streamID=%d", e.StreamID),
+				)
+			}
+
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case producer.input <- eventOrError{event: &e.RangeFeedEvent}:
+			}
+		}
+	}
+}
+
+// receiveEventsFromNode receives mux rangefeed events, and forwards them to the consumer channel.
+func (m *muxer) receiveEventsFromNode(
+	ctx context.Context, nodeID roachpb.NodeID, stream roachpb.Internal_MuxRangeFeedClient,
+) error {
+	for {
+		event, streamErr := stream.Recv()
+
+		if streamErr != nil {
+			if err := m.propagateStreamTerminationErrorToConsumers(ctx, nodeID, streamErr); err != nil {
```

Comment:
Something seems off about the error handling here. Doesn't this mean that, for example, if a single range splits or merges, we're now tearing down _all_ of the rangefeeds on that node for this client? That seems very excessive, since range split/merges and lease transfers are frequent events, and restarting these rangefeeds can be costly.

Shouldn't we pass those kinds of errors via `MuxRangeFeedEvent` so that they are only propagated to the single relevant streamID consumer? Of course, transport-level errors must still tear down all of the muxed streams.

Reply:
You're right, my bad -- didn't catch that they were already propagated as events.